### PR TITLE
Add options to check or uncheck all people in reply to modal

### PR DESCRIPTION
**Pull request by othomation**
Hello,

I tried to add a "Select all" / "Unselect all" option for reply to modal.

It may be tidious in a way that I didn't find a unique identifier for this specific modal, so I tried with nested childNodes, you will see in the code.

It does work, tho.
I am open to any improvments!

A gif : 

Gif of the select / unselect all options

### CODE REVIEW

File: src/features/addUncheckAllToReplyToModal.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  /**
+   * Listen to modal that are appended to the modal container
+   *
+   * Idea is to check if it's the right modal, since it doesn't appear to have a uuid or other consistant identifier
+   */
+  const observer = new MutationObserver((mutationsList, observer) => {
+    if (mutationsList[0].addedNodes.length > 0) {
+      const addedNode = mutationsList[0].addedNodes[0];
+      const isReplyToModal =
+        document.querySelector('.js-reply-checkbox') instanceof HTMLInputElement;
```
Comment (eramdam) [inline src/features/addUncheckAllToReplyToModal.ts:21]: Nitpick: you could also do something like
```suggestion
      const isReplyToModal = Boolean(document.querySelector('input.js-reply-checkbox'))
```

File: src/features/addUncheckAllToReplyToModal.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+import {makeBTDModule} from '../types/btdCommonTypes';
+
+enum LABELS {
+  SELECT_ALL = 'Select All',
+  UNSELECT_ALL = 'Unselect All',
+}
+
+export const addUncheckAllToReplyToModal = makeBTDModule(({TD}) => {
+  const jsModalsContainer = document.querySelector('.js-modals-container') as HTMLDivElement;
```
Comment (eramdam) [inline src/features/addUncheckAllToReplyToModal.ts:11]: `document.querySelector` is generic so you should do this instead. Your type casting here removes the `| undefined` that is usually there so you could end up in a situation where `.js-modals-container` is not defined yet TypeScript won't bat an eye.

```suggestion
  const jsModalsContainer = document.querySelector<HTMLDivElement>('.js-modals-container');
```

File: src/features/addUncheckAllToReplyToModal.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        checkbox.hidden = true;
+        checkbox.id = 'uncheck_all_reply_to_modal_checkbox';
+
+        checkbox.addEventListener('change', (e) => {
+          const selectAll = checkbox.checked;
+          label.textContent = selectAll ? LABELS.UNSELECT_ALL : LABELS.SELECT_ALL;
+          Array.from(
+            document.querySelectorAll(
+              '.js-reply-checkbox:not([disabled])'
+            ) as NodeListOf<HTMLInputElement>
```
Comment (eramdam) [inline src/features/addUncheckAllToReplyToModal.ts:44]: Same comment as above

```suggestion
            document.querySelectorAll<HTMLInputElement>(
              '.js-reply-checkbox:not([disabled])'
            )
```

File: src/features/addUncheckAllToReplyToModal.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+   */
+  const observer = new MutationObserver((mutationsList, observer) => {
+    if (mutationsList[0].addedNodes.length > 0) {
+      const addedNode = mutationsList[0].addedNodes[0];
+      const isReplyToModal =
+        document.querySelector('.js-reply-checkbox') instanceof HTMLInputElement;
+
+      if (isReplyToModal) {
+        const containerForNewCheckbox =
+          addedNode.childNodes[1].childNodes[5].childNodes[0].childNodes[1];
```
Comment (eramdam) [inline src/features/addUncheckAllToReplyToModal.ts:25]: I like to use early return in the code to avoid increasing the indentation. In that case you could do something like

```js
if (!isReplyToModal) {
return;
}

// rest of your logic
```

File: src/features/addUncheckAllToReplyToModal.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+              if (!input.checked) input.click();
+            } else {
+              if (input.checked) input.click();
+            }
+          });
+        });
+
+        Object.assign(label.style, {
+          cursor: 'pointer',
+        });
```
Comment (eramdam) [inline src/features/addUncheckAllToReplyToModal.ts:56]: Why is that necessary? 😅 You have a CSS class applied to your label above so you could just use that, no?

File: src/features/addUncheckAllToReplyToModal.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+        const container = document.createElement('div');
+        container.append(label, checkbox);
+        containerForNewCheckbox.appendChild(container);
+      }
+    }
+  });
+
+  observer.observe(jsModalsContainer, {characterData: false, childList: true, attributes: false});
+});
```
Comment (eramdam) [inline src/features/addUncheckAllToReplyToModal.ts:66]: I realize this is unlikely but I prefer to be careful: as far as I can tell your code doesn't try to check for the presence of the label you create? Meaning that you _could_ end up in an infinite loop where your MutationObserver listens for mutations, to which your callback will react by adding a DOM element, which will trigger the callback, and so on. Could you add some checks to prevent that?

File: src/features/addUncheckAllToReplyToModal.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+      const isReplyToModal =
+        document.querySelector('.js-reply-checkbox') instanceof HTMLInputElement;
+
+      if (isReplyToModal) {
+        const containerForNewCheckbox =
+          addedNode.childNodes[1].childNodes[5].childNodes[0].childNodes[1];
+
+        const label = document.createElement('label');
+        label.textContent = LABELS.UNSELECT_ALL;
+        label.setAttribute('for', 'uncheck_all_reply_to_modal_checkbox');
```
Comment (eramdam) [inline src/features/addUncheckAllToReplyToModal.ts:29]: If we're using an enum for the text content, we might as well use a variable for that id to prevent typos in the future.

**Review by eramdam — CHANGES_REQUESTED** [review]
The approach looks good! Left a few comments that are more about TypeScript/code style stuff.

**Comment by othomation** [comment]
Hello !

Thanks for your comments and the time you took!

I'm gonna proposes changes when I have some free time;

have a good day!

**Comment by eramdam** [comment]
I'll close for inactivity. Feel free to re-open the PR if/when you resume work on it 😄

**Comment by othomation** [comment]
> I'll close for inactivity. Feel free to re-open the PR if/when you resume work on it 😄

Yes, thanks you. Sorry I didn't communicate well about! Gone through a lot recently 
Will do when I hope I will have more time :D

**Comment by eramdam** [comment]
> > I'll close for inactivity. Feel free to re-open the PR if/when you resume work on it 😄
> 
> Yes, thanks you. Sorry I didn't communicate well about! Gone through a lot recently Will do when I hope I will have more time :D

No worries :)
